### PR TITLE
Add flow descriptions

### PR DESCRIPTION
**Pull request by madkinsz**
Rebase of https://github.com/PrefectHQ/prefect/pull/7775; see that pull request for discussion and context.

Authored by @J-F-Far

Closes https://github.com/PrefectHQ/prefect/issues/7774

### Checklist

- [ ] This pull request references any related issue by including "closes `<link to issue>`"
	- If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect/issues/new/choose) first.
- [x] This pull request includes tests or only affects documentation.
- [x] This pull request includes a label categorizing the change e.g. `fix`, `feature`, `enhancement`, `docs`.

### CODE REVIEW

File: src/prefect/server/database/migrations/versions/postgresql/2022_12_05_185028_adabad235795_add_description_to_flow.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
+"""empty message
```
Comment (serinamarie) [inline src/prefect/server/database/migrations/versions/postgresql/2022_12_05_185028_adabad235795_add_description_to_flow.py:1]: 🕳️

File: src/prefect/server/database/migrations/versions/sqlite/2022_12_05_184519_dc729e5c0448_add_description_to_flow.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
+"""empty message
```
Comment (serinamarie) [inline src/prefect/server/database/migrations/versions/sqlite/2022_12_05_184519_dc729e5c0448_add_description_to_flow.py:1]:
```suggestion
"""Add flow description
```

**Review by serinamarie — APPROVED** [review]
LGTM! We'd probably just want to add the changes to the migration notes.

**Review by cicdw — CHANGES_REQUESTED** [review]
I don't think we should merge this - flows are emergent entities only, they don't have any meaningful persistence in the backend and this risks a frequently updating field that changes from run-to-run.

**Comment by zhen0** [comment]
Do you mind if I close this one @madkinsz?

**Comment by madkinsz** [comment]
We do set default deployment descriptions to the flow description https://github.com/PrefectHQ/prefect/blob/9c0440eb356483bbc938d9231fef0998c9c0e905/src/prefect/deployments.py#L794 but perhaps the description should only be exposed on the deployment if we're not going to store the flow description. We can still parse the flow's docstring for default descriptions, just in the logic for deployments.

**Comment by znicholasbrown** [comment]
I have no plans for the flow description atm but I _could_ if we want to support it.

**Comment by billpalombi** [comment]
I just discussed this with @cicdw. Flows should have descriptions. Its useful for users. However, we don't want to encourage users to create/update flow descriptions with each flow run, so lets drop the `description` parameter from the `@flow` decorator and only allow flow descriptions to be created/updated via the UI/API.

**Comment by madkinsz** [comment]
> I have no plans for the flow description atm but I could if we want to support it.

Interestingly, the deployment description is probably always an exact duplicate right now. How are we going to handle the presence of both? If we deprecate `Flow.description` client-side then I guess it'd be empty until a user updates it in the UI.

> However, we don't want to encourage users to create/update flow descriptions with each flow run, so lets drop the description parameter from the @flow decorator and only allow flow descriptions to be created/updated via the UI/API.

Without this pull request, there is no flow description in the backend. We can deprecate the `Flow.description` attribute and `@flow(description=...)` here but keep the backend changes. I'd like to understand how we intend to expose / display this in the UI before putting that work in though.

**Comment by billpalombi** [comment]
> the deployment description is probably always an exact duplicate right now.

I'm sorry, I didn't follow, what is it a duplicate of?

> If we deprecate Flow.description client-side then I guess it'd be empty until a user updates it in the UI.

Yes, exactly.

> I'd like to understand how we intend to expose / display this in the UI before putting that work in though.

That's totally reasonable. I can mock that up.

I think it makes sense to break this into at least two seperate steps:
1. Deprecate Flow.description client-side
2. Add a description to the UI/server-side

**Comment by madkinsz** [comment]
> I'm sorry, I didn't follow, what is it a duplicate of?

The deployment description currently defaults to the flow description which defaults to the flow function's docstring; so if we turned on flow descriptions naively we'd have the same description (the flow docstring) displayed twice. The problem with updating deployment descriptions to read the flow function's docstring by default and removing a default flow description is that then each deployment (for a given flow) will have a duplicate default description. The default value (the docstring of a flow function) is a very reasonable flow description but it makes much less sense for deployments.

**Comment by billpalombi** [comment]
Ah I see, thank you for the explainer. Yeah I agree that it doesn't make sense to use the docstring of a flow function for the deployment description. It sounds like our current setup is a bit of a hack.

**Pull request closed without merging**